### PR TITLE
CLN: Use pydantic in FmuProvider

### LINKED ISSUE #572: Use pydantic model for metadata from FmuProvider
The `FmuProvider` should provide the metadata using the `meta.FMUClassMetaData` pydantic model instead of a dictionary

Part of https://github.com/equinor/fmu-dataio/issues/457

### PR DESCRIPTION
**Pull request by tnatt**
PR to use pydantic for metadata in the FmuProvider, and closes #572 and #577

Some additional refactoring of the FmuProvider was done
- make it more visible from the init what methods are setting the internal private properties.
- refactoring the logic for parsing folder information. Issue #557 will be addressed in an upcoming PR.

### CODE REVIEW

**Comment by tnatt** [comment]
I will investigate what I did and get the tests running 😄  they were running locally...

File: src/fmu/dataio/providers/_fmu.py [hunk 1]
```diff
@@ -84,318 +97,239 @@ class FmuProvider:
...
+        warn("Case metadata does not exist, metadata will be empty!", UserWarning)
+        return None
 
+    def _get_restart_data_uuid(self) -> UUID | None:
+        """Load restart_from information"""
+        assert self._runpath is not None
         logger.debug("Detected a restart run from environment variable")
-        restart_path = Path(self._iter_path) / environ[RESTART_PATH_ENVNAME]
-        restart_iter = self._get_folderlist_from_path(restart_path)[-1]
+        restart_path = self._runpath / environ[RESTART_PATH_ENVNAME]
```
Comment (mferrera) [inline src/fmu/dataio/providers/_fmu.py:247]: Is there possibly missing exception handling here? The previous code had
```python
if not environ.get(RESTART_PATH_ENVNAME):
    return
```
prior to this dict access. If something like that is no longer needed, it's still preferable to use `.get` or `os.getenv` I think

File: src/fmu/dataio/providers/_fmu.py [hunk 2]
```diff
@@ -84,318 +97,239 @@ class FmuProvider:
...
+        )
+
+    def _get_iteration_meta(self, iter_uuid: UUID) -> meta.Iteration:
+        return meta.Iteration(
+            id=self._iter_id,
+            name=self._iter_name,
+            uuid=iter_uuid,
+            restart_from=self._get_restart_data_uuid()
+            if environ.get(RESTART_PATH_ENVNAME)
+            else None,
```
Comment (mferrera) [inline src/fmu/dataio/providers/_fmu.py:325]:
```suggestion
            restart_from=(
                self._get_restart_data_uuid()
                if environ.get(RESTART_PATH_ENVNAME)
                else None
            ),
```
I think this adds a lot to the readability

**Review by mferrera — COMMENTED** [review]
Very nice work 👍 I think there is one definite bug with the swapped realization/iteration getter, and some comments for your consideration.

I also think we should pin pydantic for the moment, get this in, and then address pydantic 2.7.0

File: src/fmu/dataio/providers/_fmu.py [hunk 3]
```diff
@@ -84,318 +97,239 @@ class FmuProvider:
...
-
-        self._provider = "ERT"
-
-        self._detect_absolute_runpath()
-        self._detect_and_update_casepath()
-        self._parse_folder_info()
-        self._read_case_metadata()
+        self._runpath = self._get_runpath_from_env()
+        self._real_id = self._get_iteration_number_from_env()
+        self._iter_id = self._get_realization_number_from_env()
```
[the remark this answers is not in the document]
Comment (tnatt) [reply src/fmu/dataio/providers/_fmu.py:129]: Oh my 😄 just checking if you are awake! This screams for a test.. will add that

File: src/fmu/dataio/providers/_fmu.py [hunk 4]
```diff
@@ -84,318 +97,239 @@ class FmuProvider:
...
+        self._casepath = self._detect_and_update_casepath()
```
[the remark this answers is not in the document]
Comment (tnatt) [reply src/fmu/dataio/providers/_fmu.py:130]: it is kind of doing two things. If proposed casepath, just validate it. If not then try to detect it from runpath. I could split that into two methods and name them accordingly to make it more clear 👍

File: src/fmu/dataio/providers/_fmu.py [hunk 5]
```diff
@@ -84,318 +97,239 @@ class FmuProvider:
...
+        return int(iter_num) if (iter_num := FmuEnv.ITERATION_NUMBER.value) else 0
 
-        For example: current is /scratch/xfield/nn/case/realization-33/iter-1
-        shall return ['scratch', 'xfield', 'nn', 'case', 'realization-33', 'iter-1']
-        """
-        return [folder for folder in str(current).split("/") if folder]
+    @staticmethod
+    def _get_realization_number_from_env() -> int:
+        """get realization number from environment variable"""
+        return int(real_num) if (real_num := FmuEnv.REALIZATION_NUMBER.value) else 0
```
[the remark this answers is not in the document]
Comment (tnatt) [reply src/fmu/dataio/providers/_fmu.py:186]: did consider it, think I landed on that I felt the function name was more descriptive than the actual code 🙂 .. but together with a comment it should be fine to have it in post init.

File: src/fmu/dataio/providers/_fmu.py [hunk 6]
```diff
@@ -84,318 +97,239 @@ class FmuProvider:
...
+    def _get_restart_data_uuid(self) -> UUID | None:
+        """Load restart_from information"""
+        assert self._runpath is not None
         logger.debug("Detected a restart run from environment variable")
-        restart_path = Path(self._iter_path) / environ[RESTART_PATH_ENVNAME]
-        restart_iter = self._get_folderlist_from_path(restart_path)[-1]
+        restart_path = self._runpath / environ[RESTART_PATH_ENVNAME]
         restart_case_metafile = (
-            restart_path / "../.." / ERT_RELATIVE_CASE_METADATA_FILE
+            restart_path.parents[1] / ERT_RELATIVE_CASE_METADATA_FILE
```
[the remark this answers is not in the document]
Comment (tnatt) [reply src/fmu/dataio/providers/_fmu.py:249]: agreed 🙂

**Pull request merged**